trunk/user/samba3/docs/manpages/vfs_default_quota.8
.\"     Title: vfs_default_quota
.\"    Author: [see the "AUTHOR" section]
.\" Generator: DocBook XSL Stylesheets v1.74.0 <http://docbook.sf.net/>
.\"      Date: 09/30/2009
.\"    Manual: System Administration tools
.\"    Source: Samba 3.0
.\"  Language: English
.\"
.TH "VFS_DEFAULT_QUOTA" "8" "09/30/2009" "Samba 3\&.0" "System Administration tools"
.\" -----------------------------------------------------------------
.\" * (re)Define some macros
.\" -----------------------------------------------------------------
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.\" toupper - uppercase a string (locale-aware)
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.de toupper
.tr aAbBcCdDeEfFgGhHiIjJkKlLmMnNoOpPqQrRsStTuUvVwWxXyYzZ
\\$*
.tr aabbccddeeffgghhiijjkkllmmnnooppqqrrssttuuvvwwxxyyzz
..
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.\" SH-xref - format a cross-reference to an SH section
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.de SH-xref
.ie n \{\
.\}
.toupper \\$*
.el \{\
\\$*
.\}
..
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.\" SH - level-one heading that works better for non-TTY output
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.de1 SH
.\" put an extra blank line of space above the head in non-TTY output
.if t \{\
.sp 1
.\}
.sp \\n[PD]u
.nr an-level 1
.set-an-margin
.nr an-prevailing-indent \\n[IN]
.fi
.in \\n[an-margin]u
.ti 0
.HTML-TAG ".NH \\n[an-level]"
.it 1 an-trap
.nr an-no-space-flag 1
.nr an-break-flag 1
\." make the size of the head bigger
.ps +3
.ft B
.ne (2v + 1u)
.ie n \{\
.\" if n (TTY output), use uppercase
.toupper \\$*
.\}
.el \{\
.nr an-break-flag 0
.\" if not n (not TTY), use normal case (not uppercase)
\\$1
.in \\n[an-margin]u
.ti 0
.\" if not n (not TTY), put a border/line under subheading
.sp -.6
\l'\n(.lu'
.\}
..
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.\" SS - level-two heading that works better for non-TTY output
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.de1 SS
.sp \\n[PD]u
.nr an-level 1
.set-an-margin
.nr an-prevailing-indent \\n[IN]
.fi
.in \\n[IN]u
.ti \\n[SN]u
.it 1 an-trap
.nr an-no-space-flag 1
.nr an-break-flag 1
.ps \\n[PS-SS]u
\." make the size of the head bigger
.ps +2
.ft B
.ne (2v + 1u)
.if \\n[.$] \&\\$*
..
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.\" BB/BE - put background/screen (filled box) around block of text
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.de BB
.if t \{\
.sp -.5
.br
.in +2n
.ll -2n
.gcolor red
.di BX
.\}
..
.de EB
.if t \{\
.if "\\$2"adjust-for-leading-newline" \{\
.sp -1
.\}
.br
.di
.in
.ll
.gcolor
.nr BW \\n(.lu-\\n(.i
.nr BH \\n(dn+.5v
.ne \\n(BHu+.5v
.ie "\\$2"adjust-for-leading-newline" \{\
\M[\\$1]\h'1n'\v'+.5v'\D'P \\n(BWu 0 0 \\n(BHu -\\n(BWu 0 0 -\\n(BHu'\M[]
.\}
.el \{\
\M[\\$1]\h'1n'\v'-.5v'\D'P \\n(BWu 0 0 \\n(BHu -\\n(BWu 0 0 -\\n(BHu'\M[]
.\}
.in 0
.sp -.5v
.nf
.BX
.in
.sp .5v
.fi
.\}
..
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.\" BM/EM - put colored marker in margin next to block of text
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.de BM
.if t \{\
.br
.ll -2n
.gcolor red
.di BX
.\}
..
.de EM
.if t \{\
.br
.di
.ll
.gcolor
.nr BH \\n(dn
.ne \\n(BHu
\M[\\$1]\D'P -.75n 0 0 \\n(BHu -(\\n[.i]u - \\n(INu - .75n) 0 0 -\\n(BHu'\M[]
.in 0
.nf
.BX
.in
.fi
.\}
..
.\" -----------------------------------------------------------------
.\" * set default formatting
.\" -----------------------------------------------------------------
.\" disable hyphenation
.nh
.\" disable justification (adjust text to left margin only)
.ad l
.\" -----------------------------------------------------------------
.\" * MAIN CONTENT STARTS HERE *
.\" -----------------------------------------------------------------
.SH "Name"
vfs_default_quota \- store default quota records for Windows clients
.SH "Synopsis"
.fam C
.HP \w'\ 'u
\FCvfs objects = default_quota\F[]
.fam
.SH "DESCRIPTION"
.PP
This VFS module is part of the
\fBsamba\fR(7)
suite\&.
.PP
Many common quota implementations only store quotas for users and groups, but do not store a default quota\&. The
\FCvfs_default_quota\F[]
module allows Samba to store default quota values which can be examined using the Windows Explorer interface\&.
.PP
By default, Samba returns NO_LIMIT the default quota and refuses to update them\&.
\FCvfs_default_quota\F[]
maps the default quota to the quota record of a user\&. By default the root user is taken because quota limits for root are typically not enforced\&.
.PP
This module is stackable\&.
.SH "OPTIONS"
.PP
default_quota:uid = UID
.RS 4
UID specifies the user ID of the quota record where the default user quota values are stored\&.
.RE
.PP
default_quota:gid = GID
.RS 4
GID specifies the group ID of the quota record where the default group quota values are stored\&.
.RE
.PP
default_quota:uid nolimit = BOOL
.RS 4
If this parameter is True, then the user whose quota record is storing the default user quota will be reported as having a quota of NO_LIMIT\&. Otherwise, the stored values will be reported\&.
.RE
.PP
default_quota:gid nolimit = BOOL
.RS 4
If this parameter is True, then the group whose quota record is storing the default group quota will be reported as having a quota of NO_LIMIT\&. Otherwise, the stored values will be reported\&.
.RE
.SH "EXAMPLES"
.PP
Store the default quota record in the quota record for the user with ID 65535 and report that user as having no quota limits:
.sp
.if n \{\
.RS 4
.\}
.fam C
.ps -1
.nf
.if t \{\
.sp -1
.\}
.BB lightgray adjust-for-leading-newline
.sp -1

        \fI[global]\fR
	\m[blue]\fBvfs objects = default_quota\fR\m[]
	\m[blue]\fBdefault_quota:uid = 65535\fR\m[]
	\m[blue]\fBdefault_quota:uid nolimit = yes\fR\m[]
.EB lightgray adjust-for-leading-newline
.if t \{\
.sp 1
.\}
.fi
.fam
.ps +1
.if n \{\
.RE
.\}
.SH "VERSION"
.PP
This man page is correct for version 3\&.0\&.25 of the Samba suite\&.
.SH "AUTHOR"
.PP
The original Samba software and related utilities were created by Andrew Tridgell\&. Samba is now developed by the Samba Team as an Open Source project similar to the way the Linux kernel is developed\&.


trunk/user/samba3/source/modules/vfs_default_quota.c
/* 
 * Store default Quotas in a specified quota record
 *
 * Copyright (C) Stefan (metze) Metzmacher 2003
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *  
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *  
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
 */

/*
 * This module allows the default quota values,
 * in the windows explorer GUI, to be stored on a samba server.
 * The problem is that linux filesystems only store quotas
 * for users and groups, but no default quotas.
 *
 * Samba returns NO_LIMIT as the default quotas by default
 * and refuses to update them.
 *
 * With this module you can store the default quotas that are reported to
 * a windows client, in the quota record of a user. By default the root user
 * is taken because quota limits for root are typically not enforced.
 *
 * This module takes 2 parametric parameters in smb.conf:
 * (the default prefix for them is 'default_quota',
 *  it can be overwrittem when you load the module in
 *  the 'vfs objects' parameter like this:
 *  vfs objects = default_quota:myprefix)
 * 
 * "<myprefix>:uid" parameter takes a integer argument,
 *     it specifies the uid of the quota record, that will be taken for
 *     storing the default USER-quotas.
 *
 *     - default value: '0' (for root user)
 *     - e.g.: default_quota:uid = 65534
 *
 * "<myprefix>:uid nolimit" parameter takes a boolean argument,
 *     it specifies if we should report the stored default quota values,
 *     also for the user record, or if you should just report NO_LIMIT
 *     to the windows client for the user specified by the "<prefix>:uid" parameter.
 *     
 *     - default value: yes (that means to report NO_LIMIT)
 *     - e.g.: default_quota:uid nolimit = no
 * 
 * "<myprefix>:gid" parameter takes a integer argument,
 *     it's just like "<prefix>:uid" but for group quotas.
 *     (NOTE: group quotas are not supported from the windows explorer!)
 *
 *     - default value: '0' (for root group)
 *     - e.g.: default_quota:gid = 65534
 *
 * "<myprefix>:gid nolimit" parameter takes a boolean argument,
 *     it's just like "<prefix>:uid nolimit" but for group quotas.
 *     (NOTE: group quotas are not supported from the windows explorer!)
 *     
 *     - default value: yes (that means to report NO_LIMIT)
 *     - e.g.: default_quota:uid nolimit = no
 *
 */

#include "includes.h"

#undef DBGC_CLASS
#define DBGC_CLASS DBGC_QUOTA

#define DEFAULT_QUOTA_NAME "default_quota"

#define DEFAULT_QUOTA_UID_DEFAULT		0
#define DEFAULT_QUOTA_UID_NOLIMIT_DEFAULT	True
#define DEFAULT_QUOTA_GID_DEFAULT		0
#define DEFAULT_QUOTA_GID_NOLIMIT_DEFAULT	True

#define DEFAULT_QUOTA_UID(handle) \
	(uid_t)lp_parm_int(SNUM((handle)->conn),DEFAULT_QUOTA_NAME,"uid",DEFAULT_QUOTA_UID_DEFAULT)

#define DEFAULT_QUOTA_UID_NOLIMIT(handle) \
	lp_parm_bool(SNUM((handle)->conn),DEFAULT_QUOTA_NAME,"uid nolimit",DEFAULT_QUOTA_UID_NOLIMIT_DEFAULT)

#define DEFAULT_QUOTA_GID(handle) \
	(gid_t)lp_parm_int(SNUM((handle)->conn),DEFAULT_QUOTA_NAME,"gid",DEFAULT_QUOTA_GID_DEFAULT)

#define DEFAULT_QUOTA_GID_NOLIMIT(handle) \
	lp_parm_bool(SNUM((handle)->conn),DEFAULT_QUOTA_NAME,"gid nolimit",DEFAULT_QUOTA_GID_NOLIMIT_DEFAULT)

static int default_quota_get_quota(vfs_handle_struct *handle, enum SMB_QUOTA_TYPE qtype, unid_t id, SMB_DISK_QUOTA *dq)
{
	int ret = -1;

	if ((ret=SMB_VFS_NEXT_GET_QUOTA(handle, qtype, id, dq))!=0) {
		return ret;
	}

	switch (qtype) {
		case SMB_USER_QUOTA_TYPE:
			/* we use id.uid == 0 for default quotas */
			if ((id.uid==DEFAULT_QUOTA_UID(handle)) &&
				DEFAULT_QUOTA_UID_NOLIMIT(handle)) {
				SMB_QUOTAS_SET_NO_LIMIT(dq);
			}
			break;
#ifdef HAVE_GROUP_QUOTA
		case SMB_GROUP_QUOTA_TYPE:
			/* we use id.gid == 0 for default quotas */
			if ((id.gid==DEFAULT_QUOTA_GID(handle)) &&
				DEFAULT_QUOTA_GID_NOLIMIT(handle)) {
				SMB_QUOTAS_SET_NO_LIMIT(dq);
			}
			break;
#endif /* HAVE_GROUP_QUOTA */
		case SMB_USER_FS_QUOTA_TYPE:
			{
				unid_t qid;
				uint32 qflags = dq->qflags;
				qid.uid = DEFAULT_QUOTA_UID(handle);
				SMB_VFS_NEXT_GET_QUOTA(handle, SMB_USER_QUOTA_TYPE, qid, dq);
				dq->qflags = qflags;
			}
			break;
#ifdef HAVE_GROUP_QUOTA
		case SMB_GROUP_FS_QUOTA_TYPE:
			{
				unid_t qid;
				uint32 qflags = dq->qflags;
				qid.gid = DEFAULT_QUOTA_GID(handle);
				SMB_VFS_NEXT_GET_QUOTA(handle, SMB_GROUP_QUOTA_TYPE, qid, dq);
				dq->qflags = qflags;
			}
			break;
#endif /* HAVE_GROUP_QUOTA */
		default:
			errno = ENOSYS;
			return -1;
			break;
	}

	return ret;
}

static int default_quota_set_quota(vfs_handle_struct *handle, enum SMB_QUOTA_TYPE qtype, unid_t id, SMB_DISK_QUOTA *dq)
{
	int ret = -1;

	switch (qtype) {
		case SMB_USER_QUOTA_TYPE:
			/* we use id.uid == 0 for default quotas */
			if ((id.uid==DEFAULT_QUOTA_UID(handle)) &&
				DEFAULT_QUOTA_UID_NOLIMIT(handle)) {
				return -1;
			}
			break;
#ifdef HAVE_GROUP_QUOTA
		case SMB_GROUP_QUOTA_TYPE:
			/* we use id.gid == 0 for default quotas */
			if ((id.gid==DEFAULT_QUOTA_GID(handle)) &&
				DEFAULT_QUOTA_GID_NOLIMIT(handle)) {
				return -1;
			}
			break;
#endif /* HAVE_GROUP_QUOTA */
		case SMB_USER_FS_QUOTA_TYPE:
			break;
#ifdef HAVE_GROUP_QUOTA
		case SMB_GROUP_FS_QUOTA_TYPE:
			break;
#endif /* HAVE_GROUP_QUOTA */
		default:
			errno = ENOSYS;
			return -1;
			break;
	}

	if ((ret=SMB_VFS_NEXT_SET_QUOTA(handle, qtype, id, dq))!=0) {
		return ret;
	}

	switch (qtype) {
		case SMB_USER_QUOTA_TYPE:
			break;
#ifdef HAVE_GROUP_QUOTA
		case SMB_GROUP_QUOTA_TYPE:
			break;
#endif /* HAVE_GROUP_QUOTA */
		case SMB_USER_FS_QUOTA_TYPE:
			{
				unid_t qid;
				qid.uid = DEFAULT_QUOTA_UID(handle);
				ret = SMB_VFS_NEXT_SET_QUOTA(handle, SMB_USER_QUOTA_TYPE, qid, dq);
			}
			break;
#ifdef HAVE_GROUP_QUOTA
		case SMB_GROUP_FS_QUOTA_TYPE:
			{
				unid_t qid;
				qid.gid = DEFAULT_QUOTA_GID(handle);
				ret = SMB_VFS_NEXT_SET_QUOTA(handle, SMB_GROUP_QUOTA_TYPE, qid, dq);
			}
			break;
#endif /* HAVE_GROUP_QUOTA */
		default:
			errno = ENOSYS;
			return -1;
			break;
	}

	return ret;
}

/* VFS operations structure */

static vfs_op_tuple default_quota_ops[] = {	
	{SMB_VFS_OP(default_quota_get_quota),	SMB_VFS_OP_GET_QUOTA,	SMB_VFS_LAYER_TRANSPARENT},
	{SMB_VFS_OP(default_quota_set_quota),	SMB_VFS_OP_SET_QUOTA,	SMB_VFS_LAYER_TRANSPARENT},

	{SMB_VFS_OP(NULL),			SMB_VFS_OP_NOOP,	SMB_VFS_LAYER_NOOP}
};

NTSTATUS vfs_default_quota_init(void);
NTSTATUS vfs_default_quota_init(void)
{
	return smb_register_vfs(SMB_VFS_INTERFACE_VERSION, DEFAULT_QUOTA_NAME, default_quota_ops);
}
